### PR TITLE
Test delayed nodes & new sync flow

### PR DESCRIPTION
**Pull request by gavraz**
* New sync flow
* Bug fixes
* The test is still not running in automation due to an awaiting bug fix. The rest of the automation tests pass.

### CODE REVIEW

File: miner/block_builder.go
Comment (antonlerner) [inline miner/block_builder.go:46]: Do we have a "StronglySynced" notion?

File: sync/syncer.go
Comment (antonlerner) [inline sync/syncer.go:71]: This is redundant since Syncer contains mesh.Mesh

File: sync/syncer.go
Comment (antonlerner) [inline sync/syncer.go:252]: add comment to explain the logic

File: sync/syncer.go
Comment (antonlerner) [inline sync/syncer.go]: add a comment about what is p2psynced

File: sync/syncer.go
Comment (antonlerner) [inline sync/syncer.go]: here as well.... explain logic

File: sync/syncer.go
Comment (antonlerner) [inline sync/syncer.go]: shouldn't we panic if this happens?

File: sync/syncer.go
Comment (antonlerner) [inline sync/syncer.go]: Add comment about this function

File: sync/syncer.go
Comment (antonlerner) [inline sync/syncer.go:308]: this function can return errors and have them printed in calling function

File: sync/worker.go
Comment (antonlerner) [inline sync/worker.go]: what should happen here?

File: timesync/clock.go
Comment (antonlerner) [inline timesync/clock.go:104]: Add UT for this

File: types/block.go
Comment (antonlerner) [inline types/block.go:221]: shouldn't we return common.Hash?

**Review by barakshani — APPROVED** [review]

File: miner/block_builder.go
Comment (gavraz) [reply miner/block_builder.go:46]: Yes, "FullySynced" tough only on paper and not in code.
The code simply continues with the term IsSynced.

File: sync/syncer.go
Comment (gavraz) [reply sync/syncer.go:71]: It is used for testing.
It is very complicated to mock the whole mesh and I didn't want to skip tests/waste a lot of time fixing it all at once.
IMO, this is a step towards having a more testable package. In addition, I would not be afraid of having multiple, smaller, interfaces. I do agree that later on, we shouldn't have the full mesh as a member. Pretty sure there is an issue on that.

File: tests/test_delayed.py
Comment (beckmani) [inline tests/test_delayed.py:19]: @gavraz Why have you duplicated this code from test_bs.py?

File: sync/syncer.go
Comment (gavraz) [reply sync/syncer.go]: Does it break our assumptions? Yes
Did something go awfully wrong? Yes
Can we continue? not really in an honest way

I think it is a panic.
@barakshani ?

File: tests/test_delayed.py
Comment (beckmani) [inline tests/test_delayed.py:81]: Why you need this?
You can simply call append on list

File: tests/test_delayed.py
Comment (beckmani) [inline tests/test_delayed.py]: Add new line

File: types/block.go
Comment (gavraz) [reply types/block.go:221]: It doesn't matter Yuval already fixed it and we will conflict on that once I'll rebase/merge (and I'll pick Yuval's version).

File: tests/test_delayed.py
Comment (beckmani) [inline tests/test_delayed.py]: Please check if you use this function. If not remove it

File: sync/syncer.go
Comment (beckmani) [inline sync/syncer.go]: use With()

File: sync/syncer.go
Comment (beckmani) [inline sync/syncer.go]: use With()

File: sync/syncer.go
Comment (beckmani) [inline sync/syncer.go]: use With()

File: sync/syncer.go
Comment (beckmani) [inline sync/syncer.go]: use With()

File: sync/syncer.go
Comment (almogdepaz) [reply sync/syncer.go]: change name please 
we always sync with p2p

File: tests/test_delayed.py
Comment (gavraz) [reply tests/test_delayed.py:19]: Please create the fixtures/functions central python file and I'll use it (in this PR or after it is ready) as we talked.

File: sync/syncer.go
Comment (barakshani) [reply sync/syncer.go]: Should be panic, however there's an edge case with the very first layer.

File: sync/syncer.go
Comment (gavraz) [reply sync/syncer.go:252]: Done

File: sync/syncer.go
Comment (gavraz) [reply sync/syncer.go]: Done

File: sync/syncer.go
Comment (gavraz) [reply sync/syncer.go]: Done

File: sync/syncer.go
Comment (gavraz) [reply sync/syncer.go]: Done

File: sync/syncer.go
Comment (gavraz) [reply sync/syncer.go:308]: Done

File: sync/worker.go
Comment (gavraz) [reply sync/worker.go]: @barakshani 
@almogdepaz 
@zalmen 
What should we do (assuming we have no "refresh peers")
Please note I think that it doesn't happen after the fix and after we use a decent timeout.
The panic only failed the UT but if that is the behavior we want I can work on that.

File: tests/test_delayed.py
Comment (gavraz) [reply tests/test_delayed.py]: Removed

File: tests/test_delayed.py
Comment (gavraz) [reply tests/test_delayed.py]: Done

File: tests/test_delayed.py
Comment (gavraz) [reply tests/test_delayed.py:81]: Not critic. Matter of style.

Comment (gavraz) [reply tests/test_delayed.py]: Not critic. Matter of style.

File: sync/syncer.go
Comment (gavraz) [reply sync/syncer.go]: Fixed

File: sync/syncer.go
Comment (gavraz) [reply sync/syncer.go]: Removed print (used for debug)

File: sync/syncer.go
Comment (gavraz) [reply sync/syncer.go]: Done

File: sync/syncer.go
Comment (gavraz) [reply sync/syncer.go]: Done

File: sync/syncer.go
Comment (gavraz) [reply sync/syncer.go]: Done

File: timesync/clock.go
Comment (gavraz) [reply timesync/clock.go:104]: Added UT

File: sync/worker.go
Comment (almogdepaz) [reply sync/worker.go]: here you should do nothing

Comment (almogdepaz) [reply sync/worker.go]: you shouldn't panic here anyway this is just the fetcher the receiving side should decide what to do on error, any way this will all be handled in the validation queue

**Pull request merged**
